sweeper/sweeper.go

```diff
@@ -0,0 +1,140 @@
+package sweeper
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/skynetlabs/pinner/database"
+	"github.com/skynetlabs/pinner/skyd"
+	"gitlab.com/NebulousLabs/errors"
+)
+
+type (
+	// Status represents the status of a sweep.
+	Status struct {
+		InProgress bool
+		Error      error
+		StartTime  time.Time
+		EndTime    time.Time
+	}
+	// Sweeper takes care of sweeping the files pinned by the local skyd server
+	// and marks them as pinned by the local server.
+	Sweeper struct {
+		staticDB         *database.DB
+		staticSchedule   *schedule
+		staticServerName string
+		staticSkydClient skyd.Client
+
+		status   Status
+		statusMu sync.Mutex
+	}
+)
+
+// New returns a new Sweeper.
+func New(db *database.DB, skydc skyd.Client, serverName string) *Sweeper {
+	return &Sweeper{
+		staticDB:         db,
+		staticSkydClient: skydc,
+		staticServerName: serverName,
+		staticSchedule:   &schedule{},
+	}
+}
+
+// Status returns a copy of the status of the current sweep.
+func (s *Sweeper) Status() Status {
+	s.statusMu.Lock()
+	st := s.status
+	s.statusMu.Unlock()
+	return st
+}
+
+// Sweep starts a new skyd sweep, unless one is already underway.
+//
+// TODO If we want to be able to uniquely identify sweeps we can issue ids
+//  for them and keep their statuses in a map. This would be the appropriate
+//  RESTful approach. I am not sure we need that because all we care about
+//  is to be able to kick off one and wait for it to end and this
+//  implementations is sufficient for that.
+func (s *Sweeper) Sweep() {
+	go s.threadedPerformSweep()
+}
+
+// UpdateSchedule schedules a new series of sweeps to be run.
+// If there are already scheduled sweeps, that schedule is cancelled (running
+// // sweeps are not interrupted) and a new schedule is established.
+func (s *Sweeper) UpdateSchedule(period time.Duration) {
+	s.staticSchedule.Update(period, s)
+}
+
+// threadedPerformSweep performs the actual sweep operation.
+func (s *Sweeper) threadedPerformSweep() {
+	s.statusMu.Lock()
+	// Double-check for parallel sweeps.
+	if s.status.InProgress {
+		s.statusMu.Unlock()
+		return
+	}
+	// Initialise the status to "a sweep is running".
+	s.status = Status{
+		InProgress: true,
+		Error:      nil,
+		StartTime:  time.Now().UTC(),
```

Comment:
nit: Why do we need the `UTC()`? We just need the server to be consistent right? If a server is not in `UTC` timezone this is going to be inaccurate.

Reply:
If we don't need it, it would make is safer since we wouldn't be reliant on a developer remembering to add the `UTC()` when getting the current time.

Reply:
It's really helpful to have all of our servers report the same time zone. This is one less thing to worry about, especially when we write things to the DB.

If we discover that this is a problem and people are forgetting about this we should add our own Time type which does it for us.

Reply:
ok, I'd add a quick comment then so that we remember the reasoning in the future. I'd put the comment on the `StartTime` in the `Status` struct. 

fine as a follow up.

Reply:
Done in #41.